### PR TITLE
Add GitHub Actions to deploy release to prod, PRs to preview environments, and destroy preview environments

### PR DESCRIPTION
**Pull request by aknysh**
## what
* Add GitHub Actions to deploy releases to prod, PRs to preview environments, and destroy the preview environments
* The actions authenticate to AWS to deploy to the S3 buckets using the provisioned GitHub OIDC providers in the corresponding AWS accounts
* In the GitHub action, check for PR label `website-no-deploy` and don't deploy to the preview environment if the label is present (allows pushing many commits to the PR w/o triggering unnecessary deployments)

## why
* Deploy to production site https://atmos.tools when a GitHub release is published
* Deploy a PR to the dev site preview environment (with the URL like https://pr-1.atmos-docs.ue2.dev.plat.cloudposse.org)
* Destroy the preview environment (the PR folder in the S3 origin bucket) when the PR is closed or merged

## deployment
* https://pr-152.atmos-docs.ue2.dev.plat.cloudposse.org/

<br>

image

<br>

image

### CODE REVIEW

File: .github/workflows/website-deploy-preview.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-region: ${{ env.AWS_REGION }}
+          role-to-assume: ${{ env.IAM_ROLE_ARN }}
+          role-session-name: ${{ env.IAM_ROLE_SESSION_NAME }}
+
+      - name: Setup Node.js 16.x
+        uses: actions/setup-node@v3
+        with:
+          node-version: 16.x
```
Comment (nitrocode) [inline .github/workflows/website-deploy-preview.yml:41]: If you use a `.nvmrc` file with `v16` you can set this line to

```suggestion
          node-version-file: '.nvmrc'
```

File: .github/workflows/website-deploy-preview.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    types:
+      - opened
+      - synchronize
+      - reopened
+    paths:
+      - "website/**"
+
+env:
+  AWS_REGION: us-east-2
+  IAM_ROLE_ARN: arn:aws:iam::068007702576:role/cplive-plat-ue2-dev-atmos-docs-github-action
```
Comment (nitrocode) [inline .github/workflows/website-deploy-preview.yml:16]: How does this work to login to awe role with ubuntu-latest runner instead of self-hosted?

File: .github/workflows/website-deploy-release.yml [hunk 3]
```diff
@@ -0,0 +1,56 @@
+name: "website-deploy-release"
```
Comment (nitrocode) [inline .github/workflows/website-deploy-release.yml:1]: I believe you can combine the release and preview workflows by using `if` conditions on the build steps

File: .github/workflows/website-deploy-preview.yml (hunk 1 above)
Comment (aknysh) [reply .github/workflows/website-deploy-preview.yml:41]: never used .nvmrc file, what the purpose of it and what issues it will solve for us? How it will be different from not using it? @nitrocode

File: .github/workflows/website-deploy-preview.yml (hunk 2 above)
Comment (aknysh) [reply .github/workflows/website-deploy-preview.yml:16]: to use the GitHub hosted runners, we need to change the scope for it from `cloudposse/infra-live` to `cloudposse` allowing all our repos to use them.
I'm not sure we want to do it or discussed that.

Using `ubuntu-latest` the or the two reasons:

It works on ubuntu-latest runner instead of self-hosted b/c we provision a GitHub OIDC provider in each account and an IAM role that the provider is assuming. It's all in one account so no cross-account permissions are required.

We are not using the hosted GitHub runners instance role/profile because the OIDC provider seems to allow better control (we create a separate IAM role for each app/service, instead of putting all the permissions (and the future ones) into the instance profile roles of the hosted GitHub runners.
Also, it is cross-account, so we will have to add the required permissions on both sides, which could be challenging to do on the GitHub runners side b/c we'll have to add a lot of diff permissions in one place.

In short, the OIDC provider with the IAM role per repo/app works exactly like the EKS OIDC provider for Kubernetes service accounts - it allow creating small roles with just the required permissions, and not having cross-account permissions

File: .github/workflows/website-deploy-release.yml (hunk 3 above)
Comment (aknysh) [reply .github/workflows/website-deploy-release.yml:1]: I thought about it, but did not want to 1) complicate the code in each workflow; 2) we always did it this way by separating the release workflow from the preview workflow (including doing it this way in Codefresh); 3) when the workflow runs, you can see in the GitHub UI what it is about

File: .github/workflows/website-deploy-preview.yml (hunk 1 above)
Comment (nitrocode) [reply .github/workflows/website-deploy-preview.yml:41]: It's for setting a default node version via `npm use`. This will have 2 benefits. One, we will be able to codify the stable node version used to generate the docs and two, the github workflow(s) can reuse the same file without having to put the version in multiple places.

Comment (aknysh) [reply .github/workflows/website-deploy-preview.yml]: added the file, it's workin ok, thanks

File: .github/workflows/website-deploy-preview.yml (hunk 2 above)
Comment (aknysh) [reply .github/workflows/website-deploy-preview.yml:16]: the OIDC provider can also we be with self-hosted runners (w/o any changes to the provider or to the iAM role) - we just need to change the GitHub scope from `cloudposse/infra-live` to `cloudposse` in the `github-runners` component (if we want to do that)

**Review by nitrocode — DISMISSED** [review]

**Comment by nitrocode** [comment]
@aknysh should this be a `no-release` since it's not a change to atmos itself ?

File: .github/workflows/website-deploy-preview.yml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        run: |
+          node --version
+          cd website
+          npm install --only=production
+          npm run build:site
+
+      - name: Copy Website to S3 Bucket PR Folder
+        run: |
+          cd website/build
+          aws sts get-caller-identity
```
Comment (korenyoni) [inline .github/workflows/website-deploy-preview.yml:61]: Is this left over from iterative development? (i.e. checking that it's assuming the correct role before issuing the `aws s3 ...` commands).

Comment (aknysh) [reply .github/workflows/website-deploy-preview.yml:61]: just for debugging - since all the infra for it is provisioned in another repo, and there are many resources involved, and if something happens, at least we'll see that the GitHub OIDC provider can't assume the correct role.
there are no secrets in there

**Review by osterman — APPROVED** [review]
